src/function/algorithm/shortest_path.cpp

```diff
@@ -0,0 +1,167 @@
+    }
+    int64_t getLength(offset_t offset) const { return visitedArray[offset]; }
+
+    void initNextFrontier() {
+        currentFrontier = std::move(nextFrontier);
+        nextFrontier = Frontier();
+    }
+};
+
+class ShortestPathLocalState : public GDSLocalState {
+public:
+    explicit ShortestPathLocalState(main::ClientContext* context) {
+        auto mm = context->getMemoryManager();
+        srcNodeIDVector = std::make_unique<ValueVector>(*LogicalType::INTERNAL_ID(), mm);
+        dstNodeIDVector = std::make_unique<ValueVector>(*LogicalType::INTERNAL_ID(), mm);
+        lengthVector = std::make_unique<ValueVector>(*LogicalType::INT64(), mm);
+        srcNodeIDVector->state = DataChunkState::getSingleValueDataChunkState();
+        dstNodeIDVector->state = DataChunkState::getSingleValueDataChunkState();
+        lengthVector->state = DataChunkState::getSingleValueDataChunkState();
+        vectors.push_back(srcNodeIDVector.get());
+        vectors.push_back(dstNodeIDVector.get());
+        vectors.push_back(lengthVector.get());
+    }
+
+    void materialize(graph::Graph* graph, const ShortestPathSourceState& sourceState,
+        FactorizedTable& table) const {
+        srcNodeIDVector->setValue<nodeID_t>(0, sourceState.sourceNodeID);
+        for (auto offset = 0u; offset < graph->getNumNodes(); ++offset) {
+            if (!sourceState.visited(offset)) {
+                continue;
+            }
+            dstNodeIDVector->setValue<nodeID_t>(0, {offset, graph->getNodeTableID()});
+            lengthVector->setValue<int64_t>(0, sourceState.getLength(offset));
+            table.append(vectors);
+        }
+    }
+
+private:
+    std::unique_ptr<ValueVector> srcNodeIDVector;
+    std::unique_ptr<ValueVector> dstNodeIDVector;
+    std::unique_ptr<ValueVector> lengthVector;
+    std::vector<ValueVector*> vectors;
+};
+
+class ShortestPath final : public GDSAlgorithm {
+public:
+    ShortestPath() = default;
+    ShortestPath(const ShortestPath& other) : GDSAlgorithm{other} {}
+
+    std::vector<common::LogicalTypeID> getParamTypeIDs() const override {
+        return {LogicalTypeID::ANY, LogicalTypeID::INT64};
+    }
+
+    std::vector<std::string> getResultColumnNames() const override {
+        return {"src", "dst", "length"};
+    }
+
+    std::vector<common::LogicalType> getResultColumnTypes() const override {
+        return {*LogicalType::INTERNAL_ID(), *LogicalType::INTERNAL_ID(), *LogicalType::INT64()};
+    }
+
+    void bind(const binder::expression_vector& params) override {
+        ExpressionUtil::validateExpressionType(*params[1], ExpressionType::LITERAL);
+        auto upperBound = params[1]->constCast<LiteralExpression>().getValue().getValue<int64_t>();
+        bindData = std::make_unique<ShortestPathBindData>(upperBound);
+    }
+
+    void initLocalState(main::ClientContext* context) override {
+        localState = std::make_unique<ShortestPathLocalState>(context);
+    }
+
+    void exec() override {
```

Comment:
Ditto that I'm skipping through these algorithm implementations, assuming they'll be rewritten.